### ISSUE TITLE
Support attachment data type on PostgreSQL (and other external datasources)

### ISSUE THREAD

**Issue by firdapaz** [Budibase/budibase#4302]
**Describe the bug**
Hi, I have image column that have **_bytea_** datatype on my postgresql. but when i want to make fields attachment it is always show error "This field setting is wrong data type for this component" What data type i should use for image on my postgresql?

**Screenshots**
<img width="464" alt="image" src="https://user-images.githubusercontent.com/62045409/152283856-8f8b7c72-c421-491a-b56d-4c4dfa88e169.png">

**Comment by mike12345567** [comment]
Hi @firdapaz - currently attachment fields are not supported for anything other than the internal database - this is something we may look at in the future..

**Comment by mituBarua** [comment]
> Hi @firdapaz - currently attachment fields are not supported for anything other than the internal database - this is something we may look at in the future.

can you explain what do you mean by "internal database" ??

**Comment by mike12345567** [comment]
Hi @mituBarua, by internal database I mean the internal tables that can be created within a Budibase app without the need for an external datasource.

**Comment by mituBarua** [comment]
if i use mysql db ,so how can i attach file?

**Comment by mike12345567** [comment]
@mituBarua currently we do not support attachments in any other database, it will be possible to use the "S3 File Upload" component to upload files to the built in MinIO instance, or to some other S3 based environment, but we do not currently support the file/binary column types of any external datasources.

**Comment by mituBarua** [comment]
ok thanks a lot , i need another query, How can i implement google map getting current location or any other location? Please help regarding this issue also

**Comment by mituBarua** [comment]
![Screenshot from 2022-02-07 19-44-07](https://user-images.githubusercontent.com/64905702/152800692-91ee4b4d-2221-4e17-aca5-58286990b9ee.png)

i don't understand why S3 File Upload is not showing in my form dropdown.

**Comment by aptkingston** [comment]
> ![Screenshot from 2022-02-07 19-44-07](https://user-images.githubusercontent.com/64905702/152800692-91ee4b4d-2221-4e17-aca5-58286990b9ee.png)
> 
> i don't understand why S3 File Upload is not showing in my form dropdown.

Hi @mituBarua. Could you check what Budibase version you're using? We only released the S3 upload component recently, so you'll need to update your installation to the latest version, and then also update your app to the latest Budibase client version when that is done. The app can be updated using the arrows icon in the top right of the header, and it will be blue whenever an upgrade is available after you have updated your installation version first.

**Comment by mituBarua** [comment]
I am using 1.0.46 version of budibase still it is not showing  s3 upload.
Please help

**Comment by efefirda** [comment]
Hi, i already use s3 but my s3 Datasource is empty even i insert inside data provider
![image](https://user-images.githubusercontent.com/31679541/155096268-450d3345-3376-43bc-bec0-d1a8c00ef753.png)

**Comment by aderici** [comment]
I'm using mysql 8.0 which supports json data type. I really would like to use attachments (not aws s3) with my mysql db. Are there any workarounds for this? I should be able to grab the generated json from attachment component and save it to mysql but i can not set it back to the component because it complains data type for component is incompatible. But it actually is.

**Comment by fizitec** [comment]
Please add attachment fields to external databse

**Comment by ecxs** [comment]
this would be most useful for me too

**Comment by shira952** [comment]
+1 this, either add support for attachment upload to server directory, or allow to add external object db other than AWS (like access to build-in MinIO)

**Comment by TheFunkyBiskit** [comment]
+1 this. This is a tremendous limitation of Budibase if you're migrating from an existing MySQL database.

**Comment by bjornclauw** [comment]
I'm using this workaround:

### 1. Add an attachment component to a form
![image](https://user-images.githubusercontent.com/32811735/222931203-26ac4b88-1614-47ea-ace1-011efc4d3261.png)
(the name for the field value can be anything)

When you see the preview appearing, the file is already uploaded to the internal minio_DB.
![image](https://user-images.githubusercontent.com/32811735/222931351-8a679052-2338-4b8e-8357-5319439dda2f.png)

### 2. Getting the link (2 possibilities)
![image](https://user-images.githubusercontent.com/32811735/222931841-c34edc4f-d91c-4cc4-8c4a-f7ff6c85cdbc.png)

#### - 2.1 `{{ New Form.Fields.attachment.[0].key }}`

-> this gets you something like: 
key=`app_4c89fb50ab18fc30fba9895be75647cd/attachments/75f45a0a-4f89-5738-a016-02f215e930e1.png`

So url to acces this file: BASEURL/prod-budi-app-assets/key 
example: www.mybdb.com/prod-budi-app-assets/75f45a0a-4f89-5738-a016-02f215e930e1.png



#### - 2.2 `{{ New Form.Fields.attachment.[0].url }}`

-> this gets you something like: 
URL= `/files/signed/prod-budi-app-assets/app_4c89cdab18fc30fba9895bfb50e75647/attachments/75f45a0a-4f89-5738-a016-02f215e930e1.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=supersecret%2F20230304%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230304T221406Z&X-Amz-Expires=3600&X-Amz-Signature=supersecretsignature&X-Amz-SignedHeaders=host`

So url to acces this file: BASEURL/URL
example: https://www.mybdb.com/files/signed/prod-budi-app-assets/app_4c89fb50e18fc30fba9895b75647cdab/attachments/75f45a0a-5738-4f89-a016-02f215e930e1.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=secret%2F20230304%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230304T221406Z&X-Amz-Expires=3600&X-Amz-Signature=supersecret&X-Amz-SignedHeaders=host

**REMARK:**
- for method 2.1 you need to login to your minio DB and set the bucket `prod-budi-app-assets` to PUBLIC, otherwise you will get an error when you try to acces the file

- for method 2.2 this is not neccesary, however i am not using it this way and I am unsure if this link will expire and render it useless. Maybe somebody with some S3 experience can answer this one?

### 3. Store value
Use a button to write the value to your database of choice (you can even iterate and send multiple attachments).
Clear the form after storing the value(s).

### 4. Delete attachment
I am using the budibase API from inside budibase (REST datasource).

#### - 4.1 get your api key from inside the builder
![image](https://user-images.githubusercontent.com/32811735/222932387-13ef5dfc-1775-427e-88f2-9c3c8565bad8.png)

#### - 4.2 create a new REST query
![image](https://user-images.githubusercontent.com/32811735/222932546-2334f836-1bd5-435b-843c-a8a54608fb37.png)
for the BASEURL you can use http://serverip:port or https://yourbudibase.com

#### - 4.3 add your budibase API key as a header
![image](https://user-images.githubusercontent.com/32811735/222932620-02eb75c2-2d65-474f-b5bc-6204fbe755da.png)

#### - 4.4 setup the POST body
![image](https://user-images.githubusercontent.com/32811735/222932693-4bef5c12-dbef-4e53-92d4-745472b1d7d6.png)

This can for example be used when you delete a row from your attachments table (you can also delete the file from your mino DB).
NOTE: the value we need to pass = bucketname/folder/filename, which happens to be the same value as **2.1**

**Comment by eizemazal** [comment]
> it will be possible to use the "S3 File Upload" component to upload files to the built in MinIO instance, or to some other S3 based environment

This is not true as of 2.9.30. I installed Budibase from DigitalOcean marketplace, and also added a space and a postgres cluster. The S3 upload component says it will not work with custom S3 endpoints, thus eliminating for me the possibility to use DO spaces and any external MinIO instances that I could roll out on my own on K8s or bare metal instead of DO space. This sounds strange to me, because I myself implemented S3 uploads to different environments, both to DO spaces and MinIO, using Javascript and Python. Even though there are differences in API (there are S3 bucket policies in DO those Minio does not support), these environments are very similar, and I do not understand, why you limit this component to Amazon S3.

I also did not find how to use Attachment control in my environment.

Do I understand it right that if I am using an external relational database (posgres or mysql) together with anything else than Amazon S3, I am out of luck with file uploads altogether?
